x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -227,54 +318,98 @@ public Automaton allowedActionsMatcher(String index) {
         return unmodifiableMap(indexPermissions);
     }
 
+    private static boolean isIndexPattern(String indexPattern) {
+        return indexPattern.startsWith("/") || indexPattern.contains("*") || indexPattern.contains("?");
+    }
+
+    private static boolean isSystemIndex(String index) {
+        return SystemIndicesNames.indexNames().contains(index);
+    }
+
     public static class Group {
         private final IndexPrivilege privilege;
-        private final Predicate<String> actionMatcher;
         private final String[] indices;
         private final Predicate<String> indexNameMatcher;
-
-        public FieldPermissions getFieldPermissions() {
-            return fieldPermissions;
-        }
-
         private final FieldPermissions fieldPermissions;
         private final Set<BytesReference> query;
+        private final BiPredicate<String, String> implicitlyAuthorizeMonitorSystemIndices;
 
         public Group(IndexPrivilege privilege, FieldPermissions fieldPermissions, @Nullable Set<BytesReference> query, String... indices) {
             assert indices.length != 0;
             this.privilege = privilege;
-            this.actionMatcher = privilege.predicate();
             this.indices = indices;
-            this.indexNameMatcher = indexMatcher(Arrays.asList(indices));
+            this.indexNameMatcher = indexMatcherPredicate(Arrays.asList(indices));
             this.fieldPermissions = Objects.requireNonNull(fieldPermissions);
             this.query = query;
+            final Predicate<String> indicesPatternsPredicate = indicesPatternsPredicate(Arrays.asList(indices));
+            this.implicitlyAuthorizeMonitorSystemIndices = (action, index) -> {
+                return SystemIndicesNames.indexNames().contains(index) && IndexPrivilege.MONITOR.predicate().test(action)
+                        && indicesPatternsPredicate.test(index) && privilege.predicate().test(action);
```

Comment:
if I understand correctly, we need the indicesPatternsPredicate to account for the fact that the indexNameMatcher will not grant access to the security indices if they were declared as a wildcard?

Reply:
Corect.

This replaces the
```
        if (!indicesAccessControl.isGranted()) {
            throw denial(auditId, authentication, action, request, permission.names());
        } else if (hasSecurityIndexAccess(indicesAccessControl)
            && MONITOR_INDEX_PREDICATE.test(action) == false
            && isSuperuser(authentication.getUser()) == false) {
            // only the XPackUser is allowed to work with this index, but we should allow indices monitoring actions through for debugging
            // purposes. These monitor requests also sometimes resolve indices concretely and then requests them
            logger.debug("user [{}] attempted to directly perform [{}] against the security index [{}]",
                authentication.getUser().principal(), action, SecurityIndexManager.SECURITY_INDEX_NAME);
            throw denial(auditId, authentication, action, request, permission.names());
        } else {
```
block in `AuthorizationService.java` .